experiments/pop_test.go

```diff
@@ -0,0 +1,255 @@
+
+	// add "should_sign(#authority, dataID, alg, pubkey)" fact requesting the client to sign the data
+	// with specified alg and the matching private key
+	builder.AddAuthorityFact(biscuit.Fact{Predicate: biscuit.Predicate{
+		Name: "should_sign",
+		IDs: []biscuit.Atom{
+			biscuit.Integer(0),
+			biscuit.Symbol("ed25519"),
+			biscuit.Bytes(pubkey),
+		},
+	}})
+
+	staticCtx := []byte("biscuit-pop-v0")
+	challenge := make([]byte, 32)
+	_, err := rng.Read(challenge)
+	require.NoError(t, err)
+
+	// add "data(#authority, dataID, content)" fact holding the data to be signed
+	builder.AddAuthorityFact(biscuit.Fact{Predicate: biscuit.Predicate{
+		Name: "data",
+		IDs: []biscuit.Atom{
+			biscuit.Integer(0),
+			biscuit.Bytes(append(staticCtx, challenge...)),
+		},
+	}})
+
+	// This caveat requires every "should_sign" fact to have a matching "valid_signature" fact,
+	// that can only provided by the verifier (due to the ambient tag)
+	// *valid(0?)<- should_sign(0?, 1?, 2?), valid_signature(#ambient, $0, $1, $2)
+	builder.AddAuthorityCaveat(biscuit.Rule{
+		Head: biscuit.Predicate{Name: "valid", IDs: []biscuit.Atom{biscuit.Variable(0)}},
+		Body: []biscuit.Predicate{
+			{Name: "should_sign", IDs: []biscuit.Atom{biscuit.SymbolAuthority, biscuit.Variable(0), biscuit.Variable(1), biscuit.Variable(2)}},
+			{Name: "valid_signature", IDs: []biscuit.Atom{biscuit.Symbol("ambient"), biscuit.Variable(0), biscuit.Variable(1), biscuit.Variable(2)}},
+		},
+	})
+
+	b, err := builder.Build()
+	require.NoError(t, err)
+
+	t.Logf("server generated biscuit:\n%s", b.String())
+
+	s, err := b.Serialize()
+	require.NoError(t, err)
+	return s, serverKey.Public()
+}
+
+func clientSign(t *testing.T, rootPubkey sig.PublicKey, pubkey ed25519.PublicKey, privkey ed25519.PrivateKey, b []byte) []byte {
+	token, err := biscuit.Unmarshal(b)
+	require.NoError(t, err)
+
+	verifier, err := token.Verify(rootPubkey)
+	require.NoError(t, err)
+
+	t.Logf("clientSign world:\n%s", verifier.PrintWorld())
+
+	// This query returns to_sign(dataID, alg, data) facts which require a signature with a private key matching pubkey.
+	// in this example: [to_sign(0, "ed25519", "hex:7369676e2074686973")]
+	toSign, err := verifier.Query(biscuit.Rule{
+		Head: biscuit.Predicate{Name: "to_sign", IDs: []biscuit.Atom{biscuit.Variable(0), biscuit.Variable(1), biscuit.Variable(2)}},
+		Body: []biscuit.Predicate{
+			{Name: "should_sign", IDs: []biscuit.Atom{biscuit.SymbolAuthority, biscuit.Variable(0), biscuit.Variable(1), biscuit.Bytes(pubkey)}},
+			{Name: "data", IDs: []biscuit.Atom{biscuit.SymbolAuthority, biscuit.Variable(0), biscuit.Variable(2)}},
+		},
+	})
+	require.NoError(t, err)
+	t.Logf("toSign:\n%#v", toSign)
+	require.Equal(t, 1, len(toSign))
+
+	// Extract data from the fact
+	dataID, ok := toSign[0].IDs[0].(biscuit.Integer)
+	require.True(t, ok)
+	alg, ok := toSign[0].IDs[1].(biscuit.Symbol)
+	require.True(t, ok)
+	data, ok := toSign[0].IDs[2].(biscuit.Bytes)
+	require.True(t, ok)
```

Comment:
We need to confirm that the data we're signing has the static context bytes to prevent key misuse, otherwise we could inadvertantly sign something we didn't intend to.